capstone/src/main/java/com/google/sps/servlets/ChallengeServlet.java

```diff
@@ -25,22 +23,54 @@
   public void doGet(String userId, HttpServletRequest request, HttpServletResponse response)
       throws IOException {
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    Query query = new Query("Challenge").addSort("timestamp", SortDirection.DESCENDING);
-    PreparedQuery results = datastore.prepare(query);
-    Challenge challenge = null;
-    ChallengeResponse challengeResponse = null;
-    // Check if there are challenges in database
-    if (Iterables.size(results.asIterable()) > 0) {
-      // Get most recent challenge in database
-      Entity entity = results.asIterable().iterator().next();
-      challenge = Challenge.fromEntity(entity);
-      // Gets whether current user has completed challenge
-      boolean hasUserCompleted = challenge.getHasUserCompleted(userId);
-      challengeResponse = new ChallengeResponse(challenge, hasUserCompleted);
-    }
+    Entity groupEntity = Group.getGroupEntity(request, response, datastore);
+    ChallengeResponse challengeResponse =
+        this.buildChallengeResponse(groupEntity, userId, datastore, response);
     ServletHelper.write(response, challengeResponse, "application/json");
   }
 
+  private ChallengeResponse buildChallengeResponse(
+      Entity groupEntity, String userId, DatastoreService datastore, HttpServletResponse response)
+      throws IOException {
+    List<Long> challengeIds =
+        (groupEntity.getProperty("challenges") == null)
+            ? new ArrayList<Long>()
+            : (ArrayList<Long>) groupEntity.getProperty("challenges");
+    Challenge challenge = this.getMostRecentChallenge(challengeIds, datastore, response);
+    if (challenge == null) {
+      return null;
+    }
+    boolean hasUserCompleted = challenge.getHasUserCompleted(userId);
+    return new ChallengeResponse(challenge, hasUserCompleted);
+  }
+
+  private Challenge getMostRecentChallenge(
+      List<Long> challengeIds, DatastoreService datastore, HttpServletResponse response)
+      throws IOException {
+    Challenge newestChallenge = null;
+    Entity newestChallengeEntity = null;
+    Long newestTimestamp = 0L;
+
+    for (Long challengeId : challengeIds) {
+      // get newest challenge
+      Entity entity = ServletHelper.getEntityFromId(response, challengeId, datastore, "Challenge");
+      Long timestamp = (Long) entity.getProperty("timestamp");
+      if (timestamp > newestTimestamp) {
+        newestTimestamp = timestamp;
+        newestChallengeEntity = entity;
+      }
+    }
```

Comment:
Looking at `updateChallengesList` it looks like you're already maintaining an order to chellengeIds based on recency. Can this order be used to guarantee that any call to `getMostRecentChallenge` only checks AT MOST all challenges which have been invalidated since the last time `getMostRecentChallenge` was called?

Reply:
changed it so that now I get the challengeId that is at the end of the arraylist since that should be the most recent one